src/MPI.jl

```diff
@@ -4,21 +4,31 @@ module MPI
 
 using Compat
 
-const depfile = joinpath(dirname(@__FILE__), "..", "deps", "src", "compile-time.jl")
-isfile(depfile) || error("MPI not properly installed. Please run Pkg.build(\"MPI\")")
+@windows_only begin
+	const depfile = "win_mpiconstants.jl"
+end
+
+@unix_only begin
+	const depfile = joinpath(dirname(@__FILE__), "..", "deps", "src", "compile-time.jl")
+	isfile(depfile) || error("MPI not properly installed. Please run Pkg.build(\"MPI\")")
+end
+
 include(depfile)
+
 include("mpi-base.jl")
 include("cman.jl")
 
 function __init__()
-    # need to open libmpi with RTLD_GLOBAL flag for Linux, before any ccall
-    # cannot use RTLD_DEEPBIND; this leads to segfaults at least on Ubuntu 15.10
-    @eval const libmpi_handle =
-        Libdl.dlopen(libmpi, Libdl.RTLD_LAZY | Libdl.RTLD_GLOBAL)
-
-    # look up all symbols ahead of time
-    for (jname, fname) in _mpi_functions
-        eval(:(const $jname = Libdl.dlsym(libmpi_handle, $fname)))
+    @unix_only begin
+        # need to open libmpi with RTLD_GLOBAL flag for Linux, before any ccall
+        # cannot use RTLD_DEEPBIND; this leads to segfaults at least on Ubuntu 15.10
+        @eval const libmpi_handle =
+            Libdl.dlopen(libmpi, Libdl.RTLD_LAZY | Libdl.RTLD_GLOBAL)
+
+        # look up all symbols ahead of time
+        for (jname, fname) in _mpi_functions
+            eval(:(const $jname = Libdl.dlsym(libmpi_handle, $fname)))
```

Comment:
> It seems that, for Windows, you instead use a (symbol, library name) pair, which will perform the lookup at run time, every time an MPI function is called.

That is not my understanding of how ``ccall`` works. If one passes a tuple as the first argument to ``ccall``, it *has* to be a constant expression, and can thus be entirely evaluated at compile time. So there should be no lookup at run time happening at all. In fact, the way MPI.jl right now calls ``ccall`` is the type of call where the first argument is actually evaluated at runtime, i.e. when one passes a function pointer as the first argument to ``ccall`` (i.e. a [indirect call](http://docs.julialang.org/en/release-0.4/manual/calling-c-and-fortran-code/?highlight=dlopen#indirect-calls)), whereas I'm using the standard way to call ``ccall`` for the windows case.

>  have heard that Julia is supposed to cache the lookups, and perform them only once

The way it is supposed to work is that the first argument is evaluated at compile time because it is constant. I don't think there is anything going on like do the lookup when called the first time, then cache something, it should all be resolved at compile time.

> but when I measured this, this wasn't the case.

If things don't work the way I described them here we should really open a bug in the main julia repo, in that case things would work very differently from how they are described in the documentation.